Authorization

stephanebastian opened the pull request:
Hello @pmlopes , 
As discussed on Gitter and Google group, here is a PR to split Authentication and Authorization.
This PR adds the interface Authorization along with implementations. It also adds the new method `Set<Authorization> User.authorizations()`

When this PR is incorporated, I will then move forward and update all Vertx-Auth projects to use the new stuff (maybe leverage the class UserImpl everywhere instead of providing their own implementation each time)

When done I will then provide another PR to update the AuthHandler to account for the new Authorization mechanism

Could you please review ? 
Thanks

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/AndAuthorizationConverter.java:
```diff
@@ -0,0 +1,42 @@
+package io.vertx.ext.auth;
```
this should be in a `impl` package I think, as it's some internal converter code not really public api.

stephanebastian replied:
Yep, I will moved all converters to 'impl'

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/Authorization.java:
```diff
@@ -0,0 +1,52 @@
+	 * @return
+	 */
+	@GenIgnore
+	boolean implies(Authorization authorization);
```
personally i find this method name confusing, i guess other alternatives such:

* attest
* verify

would be better, but it's just IMO

stephanebastian replied:
yes, verify seems better

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/AuthorizationContext.java:
```diff
@@ -0,0 +1,26 @@
+ *
+ */
+@VertxGen
+public interface AuthorizationContext {
```
I've some concerns this this interface. It assumes that authZ is http based, perhaps that should be part of vertx-web. In vertx-auth there should not be such assumption.

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/AuthorizationConverter.java:
```diff
@@ -0,0 +1,58 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/NotAuthorizationConverter.java:
```diff
@@ -0,0 +1,31 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/OrAuthorizationConverter.java:
```diff
@@ -0,0 +1,42 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/PermissionBasedAuthorizationConverter.java:
```diff
@@ -0,0 +1,39 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/RoleBasedAuthorizationConverter.java:
```diff
@@ -0,0 +1,40 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/WildcardPermissionBasedAuthorizationConverter.java:
```diff
@@ -0,0 +1,39 @@
+package io.vertx.ext.auth;
```
move to `impl`

pmlopes commented on vertx-auth-common/src/test/java/io/vertx/ext/auth/ChainAuthTest.java:
```diff
@@ -1,11 +1,10 @@
 package io.vertx.ext.auth;
```
I guess this whole file should be deprecated as `and`/`or` operations are handled by the `Authorization` interfaces right?

stephanebastian replied:
Yes it should be deprecated

pmlopes commented on vertx-auth-htdigest/src/main/java/io/vertx/ext/auth/htdigest/impl/HtdigestUser.java:
```diff
@@ -56,4 +60,9 @@ public JsonObject principal() {
   public void setAuthProvider(AuthProvider authProvider) {
   }
+
+  @Override
```
perhaps create a default method in the interface which avoids having to patch all the sub projects?

stephanebastian replied:
yep, will do

stephanebastian commented:
Paulo, 
Thanks for the feedback.
I believe this PR now contains fixes for all your comments

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/Authorization.java:
```diff
@@ -0,0 +1,58 @@
+   * @param context
+   * @return
+   */
+  @GenIgnore
```
This is a critical interface to the API, but this means it can't be used by the polyglot languages? I am guessing this would not really need to be ignored as the parameter is also annotated with `@VertxGen`

pmlopes commented on vertx-auth-common/src/main/java/io/vertx/ext/auth/Authorization.java:
```diff
@@ -0,0 +1,58 @@
+   * @param authorization
+   * @return
+   */
+  @GenIgnore
```
Same as above

stephanebastian commented:
I removed GenIgnore and added copyright headers expected for the file WildcardExpression which is a file I adapted from Apache Shiro. So I'm not the owner or initial creator of the file.
I added the original apache header. Is this ok ?

stephanebastian commented:
concerning the ECA, I have checked yesterday and everything should be fine.

francoisprunier commented:
Hi @pmlopes 

Stéphane is off at the moment. In your opinion, is this PR ready to be merged ? I'd like to get started on the next PR if everything is fine for you.

Thanks,
François
